corehq/apps/users/views/mobile/users.py

```diff
@@ -1126,6 +1128,71 @@ def get(self, request, domain, *args, **kwargs):
         )
 
 
+class DeleteCommCareUsers(BaseManageCommCareUserView):
+    urlname = 'delete_commcare_users'
+    page_title = ugettext_noop('Bulk Delete')
+    template_name = 'users/bulk_delete.html'
+
+    @property
+    def page_context(self):
+        context = self.main_context
+        context.update({
+            'bulk_upload_form': get_bulk_upload_form(),
+        })
+        return context
+
+    def post(self, request, *args, **kwargs):
+        try:
+            workbook = get_workbook(request.FILES.get('bulk_upload_file'))
+        except WorkbookJSONError as e:
+            messages.error(request, str(e))
+            return self.get(request, *args, **kwargs)
+
+        try:
+            sheet = workbook.get_worksheet()
+        except WorksheetNotFound:
+            messages.error(request, _("Workbook has no worksheets"))
+            return self.get(request, *args, **kwargs)
+
+        try:
+            usernames = [format_username(row['username'], request.domain) for row in sheet]
+        except KeyError:
+            messages.error(request, _("No users found. Please check your file contains a 'username' column."))
+            return self.get(request, *args, **kwargs)
+
+        user_docs_by_id = {doc['_id']: doc for doc in get_user_docs_by_username(usernames)}
+        usernames_not_found = set(usernames).difference(set([doc['username'] for doc in user_docs_by_id.values()]))
+        user_ids_with_forms = (
+            FormES()
+            .domain(request.domain)
+            .user_id([doc['_id'] for doc in user_docs_by_id.values()])
+            .terms_aggregation('form.meta.userID', 'user_id')
+        ).run().aggregations.user_id.keys
+
+        deleted_count = 0
+        for user_id, doc in user_docs_by_id.items():
+            if user_id not in user_ids_with_forms:
+                CommCareUser.wrap(doc).delete()
```

Comment:
there's no need to wrap this:
```suggestion
                CommCareUser.get_db().delete_doc(doc)
```

You could also do this in bulk which is easy using the `IterDB`
```python
with IterDB(db) as iter_db:
    for user_id, doc in user_docs_by_id.items():
        if user_id not in user_ids_with_forms:
            iter_db.delete(doc)

error_ids = iter_db.error_ids
```

Reply:
Thats nice, handy iteration tool.

Reply:
Though if not wrapped, the extra stuff that happens when `delete()` is called will need to be done manually.  Notably, it looks like that's how the django user is deleted:
https://github.com/dimagi/commcare-hq/blob/5361705c913595e5dc6e0498560c327544f2839e/corehq/apps/users/models.py#L1191-L1192
`clear_quickcache_for_user` and `delete_demo_restore_for_user` are also called on delete.

Reply:
Leaving this as is because I'd rather not copy in the cache clearing, etc. Will revisit this if this performs badly.